llvm/lib/Target/WebAssembly/WebAssemblyMemorySafety.cpp

```diff
@@ -227,69 +286,115 @@ bool WebAssemblyMemorySafety::runOnFunction(Function &F) {
+                        },
+                        false));
   auto SafeMallocFn = F.getParent()->getOrInsertFunction(
       "__wasm_memsafety_malloc",
       FunctionType::get(PointerType::getInt8PtrTy(Ctx),
                         {
                             Type::getInt64Ty(Ctx),
+                        },
+                        false));
+  auto SafeCallocFn = F.getParent()->getOrInsertFunction(
+      "__wasm_memsafety_calloc",
+      FunctionType::get(PointerType::getInt8PtrTy(Ctx),
+                        {
+                            Type::getInt64Ty(Ctx),
+                            Type::getInt64Ty(Ctx),
+                        },
+                        false));
+  auto SafeReallocFn = F.getParent()->getOrInsertFunction(
+      "__wasm_memsafety_realloc",
+      FunctionType::get(PointerType::getInt8PtrTy(Ctx),
+                        {
+                            Type::getInt8PtrTy(Ctx),
                             Type::getInt64Ty(Ctx),
                         },
                         false));
   auto SafeFreeFn = F.getParent()->getOrInsertFunction(
-      "__wasm_memsafety_free", FunctionType::get(Type::getVoidTy(Ctx),
-                                                 {
-                                                     Type::getInt8PtrTy(Ctx),
-                                                 },
-                                                 false));
-
-  for (auto [Kind, Call] : CallsToAllocFunctions) {
-    switch (Kind) {
-    case llvm::AllocFnKind::Alloc: {
-      // TODO: handle functions other than c malloc
-      auto *NewCall = CallInst::Create(SafeMallocFn,
-                                       {ConstantInt::get(Type::getInt64Ty(Ctx),
-                                                         /* align = */ 16),
-                                        Call->getArgOperand(0)},
-                                       Call->getName(), Call);
-      Call->replaceAllUsesWith(NewCall);
-      break;
-    }
-    case llvm::AllocFnKind::Free: {
-      auto *NewCall = CallInst::Create(SafeFreeFn, {Call->getArgOperand(0)},
-                                       Call->getName(), Call);
-      Call->replaceAllUsesWith(NewCall);
-      break;
-    }
-    default:
-      llvm_unreachable("Not yet implemented.");
+      "__wasm_memsafety_free",
+      FunctionType::get(Type::getVoidTy(Ctx),
+                        {
+                            Type::getInt8PtrTy(Ctx),
+                        },
+                        false));
+
+
+  for (auto [AllocData, Call] : CallsToAllocFunctions) {
+    switch (AllocData.kind) {
+      case FlattenedAllocKind::AlignedAlloc: {
+        // Handle functions like C's aligned_alloc
+        auto *NewCall = CallInst::Create(SafeAlignedAllocFn,
+                                        // TODO: maybe we can just use getArgOperand(0) directly, that would be cleaner
+                                        {AllocData.alignment,
```

Comment:
After collecting the alignment value before from the `Attr`, I realized maybe it's possible to just use `getArgOperand(0)`, since we expect the `Call` to have the `aligned_alloc(alignment, size)` form. What should we use here? I think my version is a little more verbose, but `getArgOperand(0)` also seems like "cheating"/it's too easy, and maybe there's some edge case I'm missing for that.